File: src/org/simbrain/network/neuron_update_rules/KuramotoRule.java
package org.simbrain.network.neuron_update_rules;

import org.simbrain.network.core.Network;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.NeuronUpdateRule;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.neuron_update_rules.interfaces.*;
import org.simbrain.util.UserParameter;
import org.simbrain.util.math.ProbDistributions.UniformDistribution;
import org.simbrain.util.math.ProbabilityDistribution;

/**
 * TODO
 * <p>
 * https://en.wikipedia.org/wiki/Kuramoto_model
 * <p>
 * K is weight N = number of fan-in nodes
 * <p>
 * TODO: Contextual increment.  Proper randomize and bounds.
 * Remove un-needed overrides.  Finish GUI.   Include time step in gui.
 */
public class KuramotoRule extends NeuronUpdateRule implements BiasedUpdateRule, DifferentiableUpdateRule, BoundedUpdateRule, ClippableUpdateRule, NoisyUpdateRule {

    /**
     * The Default upper bound.
     */
    private static final double DEFAULT_UPPER_BOUND = 1.0;

    /**
     * The Default lower bound.
     */
    private static final double DEFAULT_LOWER_BOUND = -1.0;

    /**
     * Default clipping setting.
     */
    private static final boolean DEFAULT_CLIPPING = true;

    /**
     * Natural Frequency.
     */
    @UserParameter(
        label = "Natural frequency",
        description = "todo.",
        defaultValue = "0.2", order = 1)
    public double naturalFrequency = 1;

    /**
     * Bias.
     */
    public double bias = 0;

    /**
     * Noise generator.
     */
    private ProbabilityDistribution noiseGenerator = UniformDistribution.create();

    /**
     * Add noise to the neuron.
     */
    private boolean addNoise = false;

    /**
     * Clipping.
     */
    private boolean clipping = DEFAULT_CLIPPING;

    /**
     * The upper bound of the activity if clipping is used.
     */
    private double upperBound = DEFAULT_UPPER_BOUND;

    /**
     * The lower bound of the activity if clipping is used.
     */
    private double lowerBound = DEFAULT_LOWER_BOUND;

    ///// NEW STUFF ////
    private double timeStep;

    @Override
    public void update(Neuron neuron) {

        timeStep = neuron.getNetwork().getTimeStep();

        double sum = 0;
        for (Synapse s : neuron.getFanIn()) {
            sum += s.getStrength() * Math.sin(s.getSource().getActivation() - neuron.getActivation());
        }
        double N = neuron.getFanIn().size();
        N = (N > 0) ? N : 1;
        double theta_dot = naturalFrequency + sum / N;

        double theta = neuron.getActivation() + (timeStep * theta_dot);
        theta = theta % (2 * Math.PI);

        // if (addNoise) {
        // val += noiseGenerator.getRandom();
        // }
        //
        // if (clipping) {
        // val = clip(val);
        // }

        neuron.setBuffer(theta);
    }

    @Override
    public double clip(double val) {
        if (val > getUpperBound()) {
            return getUpperBound();
        } else if (val < getLowerBound()) {
            return getLowerBound();
        } else {
            return val;
        }
    }

    @Override
    public Network.TimeType getTimeType() {
        return Network.TimeType.DISCRETE;
    }

    @Override
    public KuramotoRule deepCopy() {
        KuramotoRule kr = new KuramotoRule();
        kr.setBias(getBias());
        kr.setSlope(getSlope());
        kr.setClipped(isClipped());
        kr.setAddNoise(getAddNoise());
        kr.setUpperBound(getUpperBound());
        kr.setLowerBound(getLowerBound());
        kr.noiseGenerator = noiseGenerator.deepCopy();
        return kr;
    }

    @Override
    public void contextualIncrement(Neuron n) {
        double act = n.getActivation();
        if (act >= getUpperBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act + n.getIncrement());
            } else {
                act = act + n.getIncrement();
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    @Override
    public void contextualDecrement(Neuron n) {
        double act = n.getActivation();
        if (act <= getLowerBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act - n.getIncrement());
            } else {
                act = act - n.getIncrement();
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    @Override
    public double getDerivative(double val) {
        if (val >= getUpperBound()) {
            return 0;
        } else if (val <= getLowerBound()) {
            return 0;
        } else {
            return naturalFrequency;
        }
    }

    @Override
    public double getBias() {
        return bias;
    }

    @Override
    public void setBias(final double bias) {
        this.bias = bias;
    }

    /**
     * @param slope The slope to set.
     */
    public void setSlope(final double slope) {
        this.naturalFrequency = slope;
    }

    @Override
    public ProbabilityDistribution getNoiseGenerator() {
        return noiseGenerator;
    }

    @Override
    public void setNoiseGenerator(final ProbabilityDistribution noise) {
        this.noiseGenerator = noise;
    }

    @Override
    public boolean getAddNoise() {
        return addNoise;
    }

    @Override
    public void setAddNoise(final boolean addNoise) {
        this.addNoise = addNoise;
    }

    @Override
    public String getName() {
        return "Kuramoto";
    }

    @Override
    public double getUpperBound() {
        return upperBound;
    }

    @Override
    public double getLowerBound() {
        return lowerBound;
    }

    @Override
    public void setUpperBound(double upperBound) {
        this.upperBound = upperBound;
    }

    @Override
    public void setLowerBound(double lowerBound) {
        this.lowerBound = lowerBound;
    }

    @Override
    public boolean isClipped() {
        return clipping;
    }

    @Override
    public void setClipped(boolean clipping) {
        this.clipping = clipping;
    }

    /**
     * @return Returns the slope.
     */
    public double getSlope() {
        return naturalFrequency;
    }

    public double getNaturalFrequency() {
        return naturalFrequency;
    }

    public void setNaturalFrequency(double naturalFrequency) {
        this.naturalFrequency = naturalFrequency;
    }

}

File: src/main/kotlin/org/simbrain/network/updaterules/KuramotoRule.kt
package org.simbrain.network.updaterules

import org.simbrain.network.core.Network
import org.simbrain.network.core.Neuron
import org.simbrain.network.updaterules.interfaces.ClippedUpdateRule
import org.simbrain.network.updaterules.interfaces.DifferentiableUpdateRule
import org.simbrain.network.updaterules.interfaces.NoisyUpdateRule
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.util.UserParameter
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.UniformRealDistribution
import kotlin.math.sin

/**
 * TODO
 *
 *
 * https://en.wikipedia.org/wiki/Kuramoto_model
 *
 *
 * K is weight N = number of fan-in nodes
 *
 *
 * TODO: Contextual increment.  Proper randomize and bounds.
 * Remove un-needed overrides.  Finish GUI.   Include time step in gui.
 */
class KuramotoRule : NeuronUpdateRule<EmptyScalarData, EmptyMatrixData>(), DifferentiableUpdateRule,
    ClippedUpdateRule, NoisyUpdateRule {
    /**
     * Natural Frequency.
     */
    @UserParameter(label = "Natural frequency", description = "todo.", increment = .1, order = 1)
    var slope: Double = 1.0
        get() = field
        set(slope) {
            field = slope
        }

    /**
     * Noise generator.
     */
    override var noiseGenerator: ProbabilityDistribution = UniformRealDistribution()

    /**
     * Add noise to the neuron.
     */
    override var addNoise: Boolean = false

    /**
     * Clipping.
     */
    override var isClipped: Boolean = DEFAULT_CLIPPING

    /**
     * The upper bound of the activity if clipping is used.
     */
    override var upperBound: Double = DEFAULT_UPPER_BOUND

    /**
     * The lower bound of the activity if clipping is used.
     */
    override var lowerBound: Double = DEFAULT_LOWER_BOUND

    context(Network)
    override fun apply(neuron: Neuron, data: EmptyScalarData) {

        var sum = 0.0
        for (s in neuron.fanIn) {
            sum += s.strength * sin(s.source.activation - neuron.activation)
        }
        var N = neuron.fanIn.size.toDouble()
        N = if ((N > 0)) N else 1.0
        val theta_dot = slope + sum / N

        var theta = neuron.activation + (timeStep * theta_dot)
        theta = theta % (2 * Math.PI)

        // if (addNoise) {
        // val += noiseGenerator.nextRand();
        // }
        //
        // if (clipping) {
        // val = clip(val);
        // }
        neuron.activation = theta
    }

    override val timeType: Network.TimeType
        get() = Network.TimeType.DISCRETE

    override fun copy(): KuramotoRule {
        val kr = KuramotoRule()
        kr.slope = slope
        kr.isClipped = isClipped
        kr.addNoise = addNoise
        kr.upperBound = upperBound
        kr.lowerBound = lowerBound
        kr.noiseGenerator = noiseGenerator.copy()
        return kr
    }

    override fun getDerivative(`val`: Double): Double {
        return if (`val` >= upperBound) {
            0.0
        } else if (`val` <= lowerBound) {
            0.0
        } else {
            slope
        }
    }

    override val name: String
        get() = "Kuramoto"

    companion object {
        /**
         * The Default upper bound.
         */
        private const val DEFAULT_UPPER_BOUND = 1.0

        /**
         * The Default lower bound.
         */
        private const val DEFAULT_LOWER_BOUND = -1.0

        /**
         * Default clipping setting.
         */
        private const val DEFAULT_CLIPPING = true
    }
}